### DISCUSSION TITLE
How to verify contracts deployed to telos network using hardhat

### DISCUSSION THREAD

**Discussion by markeddolphin** [telosnetwork/teloscan#429] [Q&A] [answered]
I deployed contracts to telos testnet and tried to verify them via hardhat.
But I couldn't find api key for teloscan like in the other evm based chains.
Is it impossible to verify contracts using hardhat for telos?

**Comment by tomtomcrypto** [comment, marked as answer]
You do not need a key at all just use the hardhat deploy sourcify plugin

`npx hardhat --network testnet sourcify`

You may run into a "gateway timeout" error message, do not worry, the contract will still be verified
You can make sure it is by running the command again, you should get this message back:
`already verified: MyContract (0x2A3b40A59109a84ab82a0Ff1A3910bb595082e09), skipping.`

If you got questions like this better to contact us directly on Telegram @ https://t.me/TelosEVMDevs
Also: https://docs.telos.net/evm/smart-contracts/hardhat_sourcify

**Comment by markeddolphin** [comment]
Nice! So verifying smart contracts using sourcify is working like a dream except for one case.
There is an upgradable proxy contract inherited from openzeppelin proxy, which is being created dynamically. and verifying using sourcify doesn't work for this one.
For the other evm based chains like avalanche-c chain, I could see contract creator for the proxy contract. But for deployed contract on telos testnet, No creator address I could see.
Is there any other special trick for the proxy contract on telos?

**Comment by tomtomcrypto** [comment]
What do you mean by dynamically ? Are you using CREATE2 on chain ? See https://docs.sourcify.dev/docs/create2/

**Comment by markeddolphin** [comment]
Let me share detail code here.

```
/**
   * @dev Updates the implementation of the LendingPool, or creates the proxy
   * setting the new `pool` implementation on the first time calling it
   * @param pool The new LendingPool implementation
   **/
  function setLendingPoolImpl(address pool) external override onlyOwner {
    _updateImpl(LENDING_POOL, pool);
    emit LendingPoolUpdated(pool);
  }
 
/**
   * @dev Internal function to update the implementation of a specific proxied component of the protocol
   * - If there is no proxy registered in the given `id`, it creates the proxy setting `newAdress`
   *   as implementation and calls the initialize() function on the proxy
   * - If there is already a proxy registered, it just updates the implementation to `newAddress` and
   *   calls the initialize() function via upgradeToAndCall() in the proxy
   * @param id The id of the proxy to be updated
   * @param newAddress The address of the new implementation
   **/
  function _updateImpl(bytes32 id, address newAddress) internal {
    address payable proxyAddress = payable(_addresses[id]);

    InitializableImmutableAdminUpgradeabilityProxy proxy = InitializableImmutableAdminUpgradeabilityProxy(
        proxyAddress
      );
    bytes memory params = abi.encodeWithSignature('initialize(address)', address(this));

    if (proxyAddress == address(0)) {
      proxy = new InitializableImmutableAdminUpgradeabilityProxy(address(this));
      proxy.initialize(newAddress, params);
      _addresses[id] = address(proxy);
      emit ProxyCreated(id, address(proxy));
    } else {
      proxy.upgradeToAndCall(newAddress, params);
    }
  }
```

Actually, I am doing deploying aave v2 contracts to telos testnet.
Here, it seems it deploys lending pool implementation contract first and call `setLendingPoolImpl`. In there, it creates proxy contract when it is being called first time.

**Comment by tomtomcrypto** [comment]
We will have creation information available with 1.5 RPCs which will be out pretty soon.
In the meantime you may be able to flatten the contract to a file as if it was deployed from hardhat directly and upload that to Sourcify. Sorry not sure how to do it exactly in your case.

**Comment by markeddolphin** [comment]
Thanks for letting me know that.
May I ask last question? Sorry, but I'm gonna debug myself according to your answer.
I tried to deploy contracts to fuji testnet and I could see tx hash and creator info of proxy contract I mentioned in the last comment. But for telos testnet, With same deploy scripts, I couldn't see any of tx hash and creator address of deployed proxy contract.

https://testnet.snowtrace.io/address/0xCF5A519b6a7d758909a342DE7104923F1fC5B683
https://testnet.teloscan.io/address/0xb1c5E5D138EB826F58c1b97Dd980908bcE625255

Here If you have any idea about this issue, Plz let me know anytime. That'd be much thankful.
Cheers

**Reply by tomtomcrypto** [reply]
Yes, as I said we don't have the creation info with our current RPC setup but it will be available once we get 1.5 deployed.  You could use our preview 1.5 testnet RPC endpoint but it isn't synced currently as we are solving a bug. Will let you know once we get it back in sync.

**Comment by markeddolphin** [comment]
Much appreciate your help!
Yeah, I will look forward to listen from you again soon.
Thanks
